### ISSUE TITLE
Finalize Kube items 3.1

### ISSUE THREAD

**Issue by pweil-** [openshift/origin#5107]
Tracking issue for unmerged upstream commits a la https://github.com/openshift/origin/issues/2233

Cherry-pick candidates:

Starting with commit `4c8e6f47ec23f390978e651232b375f5f9cde3c7`

Kube PR      | Origin PR | Description                              | Justification
-------      | --------- | ---------------------------------------- | -------------
15958        | ~~#5257~~ | URL API detection                        | Needed for extensions API enablement
15232        | ~~#4997~~ | Refactor log command to be reusable      | Needed for deployment logs, build logs
15784        | ~~#5278~~ | remove_blank_proxy_headers               | 
15427        | out       | controllerversion                        | Nice to have per @deads2k, not essential
11694        | ~~#5397~~ | add-spdy-proxy-support                   | needed per @ncdc
15520        | ~~#5490~~ | Job generalized label selector | API change to JobSpec
15651        | out       | label-patch                              | nice to have, per @smarterclayton
15597        | out       | annotate-patch                           | nice to have, per @smarterclayton
15706        | ~~#5310~~ | Promote HPA and scale APIs | HPA/Scale API changes
15183        | out/#5540 | Enable Pod garbage collection | Performance impact, need to update our setting as well
15194        | ~~#5204~~ | Fix kubelet host network setup | 
15574        | ~~#5642~~ | resource quota validation       | Validation change
15646        | ~~#5643~~ | daemonset validation  | Validation change
15799        | ~~#5494~~ | podstatus crash loop backoff                               | P0 fix
15682        | out       | make_proxy_resync_longer                 | Performance impact, done partially in a41d9dac9a66bfd8afb0c2a26642d4b31cb6919f
15689        | out       | root scope componentstatus               | Not sure we need this, but it's an API change
15053        | ~~#4819~~, ~~#5338~~  | stdin once, fix attach | needed for binary build, attach fix
15733        | ~~#5556~~ | disable keepalive for liveness probes                        | Performance fix
14161        | out       | letKubectlExecFollowDashRule             | maybe per @smarterclayton 
15745        | ~~#5644~~ | reduce endpoint timeout                            | P0 fix
15739        | out       | kubectl-autoscale                        | @ncdc do we want this?
15612, 16127 | ~~#5581~~ | cadvisor/libcontainer bump                         | Performance impact. Edit: hold for https://github.com/kubernetes/kubernetes/pull/16127 instead? Edit 2: bump to cadvisor v0.19.2 instead
15845        | ~~#5555~~ | service-locator                          | Nice to have, makes service proxy more usable
14881        | out/#5339 | delta fifo / go1.5.1 fixes | should have, per @deads2k. blocked by build controller usage of delta fifo, per @bparees 
15414        | ~~#5646~~ | beta-annotations                     | annotation key changes from "experimental"
14991        | ~~#5169~~ | supplemental-groups                      | needed per @pmorie
15944        | ~~#5647~~     | daemon set controller assignment fix                            | Merged, but might need revert. Reverted in ~~#5702~~
16020        | out       | fix_request_tests                        | test fix
15983, 16080 | ~~#5480~~ | mirror pod fix  | P0 fix
14705        | ~~#5169~~ | pod-sc-inline                            | needed per @pmorie
16068        | ~~#5550~~ | increase-annotation-size-limit           | Validation change, needed for `apply`
16044        | ~~#5650~~ | store-err                                | cache store error handling
16042        | ~~#5337~~ | missing-err-handling                     | 
15997        | ~~#5648~~ | kubectl-get-hpa-minpods                  | fix NPE in kubectl
15975        | ~~#5430~~ | validate_before_create                   | Validation change
15900        | ~~#5533~~ | fix deletion of terminated pods        | Should have, per @smarterclayton 
15930        | ~~#5533~~ | deletion of terminated pods on 1.0 nodes  | @smarterclayton, is this needed?
14182        | ~~#5653~~ | prereq to serialize kubelet image pulls |
15914, 16749 | ~~#5657~~     | serialize kubelet image pulls | needed per https://github.com/openshift/origin/issues/5107#issuecomment-151514430
16025        | ~~#5649~~ | fix describer bug for HPA | nice to have
16052        | ~~#5651~~ or ~~#5652~~ | kubelet managed /etc/hosts file | must have, may prereq docker https://github.com/docker/docker/pull/14613
16137        | ~~#5535~~ | release NodePort correctly |
15352        | ~~#5169~~ | FSGroup impl |
16109        | ~~#5271~~ | Attach support for DeploymentConfig |
16174        | ~~#5652~~ | Nil check for security context | @pmorie, @pweil-, do we need this?
16266, 15705, 16369 |out?| Timeout, watch reset | maybe not, per https://github.com/openshift/origin/issues/5107#issuecomment-151523879
16112        | ?         | expose ResyncPeriod for easier composition | @deads2k, do we need this?
15961        | ~~#5525~~ | streaming subprotocol negotiation | @ncdc, do we need this?
16221        | ?         | RESTMapper KindToResource impl change, ingress->ingresses | will mean a carry patch for any pluralized resource kinds we have
16058        | ?         | revert slash syntax for `explain` |
16241        | ~~#5419~~ | Fix wsstream test |
16086        | ?         | only update `apply` annotation if already present | 
15953        | ~~#5249~~ | return errors directly from negotiation | 
16191        | ~~#5654~~ | allow getting logs from mirror pods | @smarterclayton, do we need this?
16053        | out       | GCE Cloud/Attach/Detach stability fixes | @smarterclayton, do we need this?
16277        | ~~#5673~~ | HPA last scale time fix | needed per https://github.com/openshift/origin/issues/5107#issuecomment-151514430
16286        | ~~#5400~~ | wsstream CPU hotloop |
16084        | ~~#5312~~ | All e2e tests should use NewFramework |
15461        | ~~#5067~~ | Allow exposing multi-port objects | @smarterclayton, do we need this?
16340        | ~~#5655~~ | Fix pod status update sync | @smarterclayton, do we need this?
16178, 16179 | out/#5599 | Report out of disk condition from kubelet, use the out of disk condition when scheduling | @smarterclayton, do we need this?
15809        | out       | Deployment status subresource | @derekwaynecarr/@smarterclayton, do we need this?
16054        | | Revert resource.Builder support for group/type | 
15323        | ~~#5169~~ | Support labeling of volumes to Pod's SELinux label | per @pmorie
15275, 16223 | ~~#5554~~ | Concurrency fixes in status.Manager, liveness | 
16494        | ~~#5490~~ | Remove dead pods upon stopping a job |
16082        | out | Ignore error when discovering server supported groupVersions and resources | probably don't want this for 3.1
16478        | ~~#5656~~ | daemonset shouldn't place pods on notready nodes | p0 fix, maybe
16443        | | don't add unscheduleable nodes to load balancers | p0 fix
16544        | ~~#5503~~ | add support for tailLines + follow logs |
16118        | ~~#5344~~ | spdystream bump to fix potential data loss | 
15791        | ~~#5094~~ | reconcile master service ports in controller loop | 
15236, 15562, 15555, 16032, 16033 | ~~#5551~~ | Storage fixes | 16032 reverted in 16752, in origin ~~#5672~~
16441        | ~~#5454~~ | Use RESTClient codec to encode for server |
16537        | ~~#5497~~ | Attach must only allow a tty when container supports it |
16570        | ~~#5310~~ | Fix URL parsing for proxy/redirect verbs |
16671        | ~~#5310~~ | Customize HPA Heapster service namespace/name |
16668        | ~~#5579~~ | Fix HPA cross-namespace escalation |
16711        | ~~#5620~~ | Restore error surfacing behavior for exec |
16717        | ~~#5682~~  | Ensure HPA scaleRef has valid name/kind/subresource |

Scanned through `5ff0e128ad7d73f7e6b781b5841704efbf62702c`

**Comment by markturansky** [comment]
Dynamic provisioning controller: https://github.com/kubernetes/kubernetes/pull/14537

**Comment by pmorie** [comment]
Volume Security:

Short-term carries, will come in in next rebase:

1. Pod security context skeleton: https://github.com/kubernetes/kubernetes/13939
2. Bump fsouza-dockerclient: https://github.com/kubernetes/kubernetes/pull/15127

Carries:

1. Supplemental Groups: https://github.com/kubernetes/kubernetes/pull/14911
2. FSGroup implementation: https://github.com/kubernetes/kubernetes/pull/15352
3. SELinux labeling for volumes: https://github.com/kubernetes/kubernetes/pull/15323
4. Inline fields into pod.Spec.SecurityContext https://github.com/kubernetes/kubernetes/pull/14705

**Comment by jsafrane** [comment]
Dynamic provisioning plugins:
AWS: https://github.com/kubernetes/kubernetes/pull/15603
GCE: https://github.com/kubernetes/kubernetes/pull/15604
OpenStack+Cinder: https://github.com/kubernetes/kubernetes/pull/15606

All of them already contain Dynamic provisioning controller, kubernetes/kubernetes#14537 from @markturansky. Merge kubernetes/kubernetes#14537 first and then just ignore the first commit when merging these plugins.

**Comment by jsafrane** [comment]
OpenStack/Cinder hostname fixes:
upstream: https://github.com/kubernetes/kubernetes/pull/15537
origin: https://github.com/openshift/origin/pull/5272

**Comment by smarterclayton** [comment]
kubernetes/kubernetes#16068 yes
kubernetes/kubernetes#14161 maybe
kubernetes/kubernetes#15682 already have cherrypicked

**Comment by ncdc** [comment]
15739, the autoscale command, is useful, but someone would probably need to write an `oc` wrapper for it, and QE may/may not be able to test it. I'll defer to @devop-mmcgrath @danmcp @smarterclayton on this one.

**Comment by smarterclayton** [comment]
kubernetes/kubernetes#15651 nice to have
kubernetes/kubernetes#15597 nice to have

**Comment by smarterclayton** [comment]
kubernetes/kubernetes#15900 should have

**Comment by soltysh** [comment]
Will start picking up job related PRs, once we have job working.

**Comment by smarterclayton** [comment]
@ncdc we need to cherrypick the port forward e2e flake fix (if it exists)

**Comment by ncdc** [comment]
What flake is that?

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/pull/16207 must have

**Comment by derekwaynecarr** [comment]
Two additional must haves to track:

make kubelet image pulls serialized by default
https://github.com/kubernetes/kubernetes/pull/15914
Works around docker image pull corruption until docker 1.9 by doing all image pulls in serial.  

Fixed resetting last scale time in HPA status
https://github.com/kubernetes/kubernetes/pull/16277 
Without this fix, HPA can make premature downscaling decisions.

**Comment by smarterclayton** [comment]
Consider kubernetes/kubernetes#16266 which is a more involved fix to
our watch reset problem.  I'm tempted to let it back in upstream for
now rather than trying to pull it for 3.1.

**Comment by derekwaynecarr** [comment]
Large amount of memory churn in etcd key prefix generation
https://github.com/kubernetes/kubernetes/pull/16384

**Comment by soltysh** [comment]
Good to have:
https://github.com/kubernetes/kubernetes/pull/16494
https://github.com/kubernetes/kubernetes/pull/16332

**Comment by markturansky** [comment]
Fix race condition in PV binder: https://github.com/openshift/origin/pull/5467

**Comment by jsafrane** [comment]
Fixed reusing of AWS EBS volumes: https://github.com/kubernetes/kubernetes/pull/14493
Issue + reproducer: https://github.com/kubernetes/kubernetes/issues/15073#issuecomment-151219755

**Comment by ncdc** [comment]
Must have 15961. I'll see about a cherry-pick.

**Comment by ncdc** [comment]
Cherry pick of 15961 is #5525

**Comment by soltysh** [comment]
Jobs are cherry-picked in #5490

**Comment by smarterclayton** [comment]
#16054 we need to carry this ourselves.

**Comment by smarterclayton** [comment]
Concurrency https://github.com/openshift/origin/pull/5532

**Comment by smarterclayton** [comment]
Graceful deletion fixes 15900 and 15930 in https://github.com/openshift/origin/pull/5533

**Comment by rootfs** [comment]
cherry-pick 

https://github.com/kubernetes/kubernetes/pull/15236
https://github.com/kubernetes/kubernetes/pull/15562
https://github.com/kubernetes/kubernetes/pull/15555
https://github.com/kubernetes/kubernetes/pull/16032
https://github.com/kubernetes/kubernetes/pull/16033

**Comment by smarterclayton** [comment]
Node port release in https://github.com/openshift/origin/pull/5535

**Comment by smarterclayton** [comment]
Cherrypick of 15651 and 15597 in https://github.com/openshift/origin/pull/5536

**Comment by liggitt** [comment]
@rootfs @smarterclayton, what's the priority on these:

> kubernetes/kubernetes#15236
> kubernetes/kubernetes#15562
> kubernetes/kubernetes#15555
> kubernetes/kubernetes#16032
> kubernetes/kubernetes#16033

none got cherry-picked into 1.1 upstream... are they high severity enough to warrant cherry picking into origin?

**Comment by smarterclayton** [comment]
Yes, maybe, maybe, yes, yes

**Comment by smarterclayton** [comment]
cAdvisor 15612 and 16127 https://github.com/openshift/origin/pull/5548

**Comment by smarterclayton** [comment]
16068 validation limits in https://github.com/openshift/origin/pull/5550

**Comment by smarterclayton** [comment]
Storage fixes 15236, 15562, 15555, 16032, 16033 in https://github.com/openshift/origin/pull/5551

**Comment by smarterclayton** [comment]
We absolutely need 16052 - @mrunalp @derekwaynecarr

**Comment by smarterclayton** [comment]
Kubelet liveness probing improvements 15275 16223 in https://github.com/openshift/origin/pull/5554

**Comment by smarterclayton** [comment]
15845 serivce locator in rest storage in https://github.com/openshift/origin/pull/5555

**Comment by smarterclayton** [comment]
Keep alive disabled in liveness in https://github.com/openshift/origin/pull/5556

**Comment by mrunalp** [comment]
@smarterclayton What is pending for 16052 in docker?

**Comment by smarterclayton** [comment]
We talked about carrying a patch, but i don't know the outcome.  @rhatdan?

On Oct 30, 2015, at 7:18 PM, Mrunal Patel <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> What is pending for
16052 in docker?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/5107#issuecomment-152672083>.

**Comment by markturansky** [comment]
https://github.com/openshift/origin/pull/5465 to resolve https://github.com/openshift/origin/issues/5399

**Comment by smarterclayton** [comment]
15183 is punted from 1.1 because it's too hard to debug.

**Comment by smarterclayton** [comment]
15651 and 15597 are punted from 1.1 because of lack of need.

**Comment by smarterclayton** [comment]
@markturansky are there any other storage issues besides the ones listed here?

**Comment by markturansky** [comment]
Nothing else, the service account and race fix are both in this list.

I will ask my teammates in just a few minutes about their potential list.

**Comment by pmorie** [comment]
#5599 covers kubernetes/kubernetes#16178 and kubernetes/kubernetes#16179

**Comment by smarterclayton** [comment]
@mheon this is the tracking issue for 3.1 - Kube 16052 is dependent on docker 14613 which means we need to have the patch in 1.8 and 1.9

**Comment by mheon** [comment]
@smarterclayton Looks like we are carrying docker 14612 already in both 1.8 and 1.9, so this should be set.

**Comment by ncdc** [comment]
#5601 - attempt 2 at fixing #5507, hopefully without race conditions this time

**Comment by markturansky** [comment]
whatever PR resolves this issue must be patched: https://github.com/kubernetes/kubernetes/issues/16727

**Comment by smarterclayton** [comment]
@tracyrankin

**Comment by derekwaynecarr** [comment]
Distinguish image registry unavailable and pull failure
https://github.com/kubernetes/kubernetes/pull/14182

is a prerequisite to pulling

make kubelet image pulls serialized by default
https://github.com/kubernetes/kubernetes/pull/15914

**Comment by smarterclayton** [comment]
16086 is out

**Comment by smarterclayton** [comment]
16277 is already in master

**Comment by smarterclayton** [comment]
16443 is out, but still p0

**Comment by smarterclayton** [comment]
16118 is already in

**Comment by smarterclayton** [comment]
5648 updated with latest messages.

**Comment by markturansky** [comment]
Fixes the recycler to work with the new binary checks in NFS plugin: https://github.com/openshift/origin/pull/5665

**Comment by markturansky** [comment]
Remove binary checks in NFS CanSupport:  https://github.com/openshift/origin/pull/5672

**Comment by sdodson** [comment]
> Large amount of memory churn in etcd key prefix generation
> kubernetes/kubernetes#16384

There's no PR for this yet is there?

**Comment by smarterclayton** [comment]
I *thought* that's in our code.

**Comment by smarterclayton** [comment]
It's not, I'm going to cherrypick

**Comment by smarterclayton** [comment]
AFAIK everything on this list is triaged as out, or merged.

**Comment by smarterclayton** [comment]
Was merged

**Comment by smarterclayton** [comment]
@liggitt ack on the completion of this?

**Comment by liggitt** [comment]
did we get an answer on 16443? not sure what the impact of shipping the CLI without 16054 is

**Comment by smarterclayton** [comment]
We'll just say it's unsupported.

**Comment by smarterclayton** [comment]
I'm ok kicking 16443 out

**Comment by liggitt** [comment]
table is complete in comment 0 is complete

**Comment by smarterclayton** [comment]
Thanks for the hard work to get this done guys.  If you have something that comes up, please open an issue and flag it p1 or p0 if it needs to be in 1.1.
